dfdaemon/proxy/proxy.go

```diff
@@ -314,7 +320,21 @@ func (proxy *Proxy) handleHTTPS(w http.ResponseWriter, r *http.Request) {
 	logrus.Debugln("hijack https request to", r.Host)
 
 	sConfig := new(tls.Config)
-	sConfig.Certificates = []tls.Certificate{*proxy.cert}
+	if proxy.cert.Leaf != nil && proxy.cert.Leaf.IsCA {
+		logrus.Debugf("hijack https request with CA <%s>", proxy.cert.Leaf.Subject.CommonName)
+		leafCertSpec := LeafCertSpec{
+			proxy.cert.Leaf.PublicKey,
+			proxy.cert.PrivateKey,
+			proxy.cert.Leaf.SignatureAlgorithm}
+		host, _, _ := net.SplitHostPort(r.Host)
+		sConfig.GetCertificate = func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			cConfig.ServerName = hello.ServerName
+			logrus.Debugf("Generate temporal leaf TLS cert for host <%s>", hello.ServerName)
+			return genLeafCert(proxy.cert, &leafCertSpec, hello.ServerName, host)
```

Comment:
I didn't actually benchmark it, but maybe we can cache certificates for each `(hello.ServerName, host)` pair, to avoid generating certificate for every connection?

Reply:
It’s a good idea. I’ll consider it later.

Reply:
Now I added cache support using `lru.Cache`